### PR TITLE
Adding new integration -  Aws Secrets Manager

### PR DESCRIPTION
**Pull request by ishaish103**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-hq.paloaltonetworks.local/browse/CIAC-2104

## Description
Aws Secrets manager integration

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+    response = json.loads(output)
+
+    readable_output = [{'Name': secret.get('Name', ''),
+                        'ARN': secret.get('ARN', ''),
+                        'Description': secret.get('Description', ''),
+                        'LastAccessedDate': secret.get('LastChangedDate', '')} for secret in response['SecretList'][offset:end]]
+
+    human_readable = tableToMarkdown('AWS Secrets List', readable_output)
+
+    return_results(CommandResults(
```
Comment (DeanArbel) [inline Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py:98]: Please add `outputs_key_field`.

File: Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+
+    readable_output = {'Name': response.get('Name', ''),
+                       'ARN': response.get('ARN', ''),
+                       'SecretBinary': response.get('SecretBinary', ''),
+                       'SecretString': response.get('SecretString', ''),
+                       'CreatedDate': response.get('CreatedDate', '')}
+
+    human_readable = tableToMarkdown('AWS Get Secret', readable_output)
+
+    return_results(CommandResults(
```
Comment (DeanArbel) [inline Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py:142]: Please add `outputs_key_field`.

File: Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+        outputs_prefix='AWS.SecretsManager.Secret.SecretValue',
+        outputs=response,
+        readable_output=human_readable
+    ))
+
+def aws_secrets_manager_secret_delete_command(client: AWSClient, args: Dict[str, Any]) -> CommandResults:
```
Comment (DeanArbel) [inline Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py:148]: This returns `None`.

File: Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+        if args.get('days_of_recovery'):
+            return_error('Delete command cannot be executed with both args: delete_immediately and days_of_recovery')
+        kwargs['ForceDeleteWithoutRecovery'] = args.get('delete_immediately', False) == True
+
+    response = client.delete_secret(**kwargs)
+
+    if response['ResponseMetadata']['HTTPStatusCode'] == 200:
+        demisto.results("The Secret was Deleted")
+
+def aws_secrets_manager_secret_restore_command(client: AWSClient, args: Dict[str, Any]) -> CommandResults:
```
Comment (DeanArbel) [inline Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py:172]: This returns `None`.

File: Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+def aws_secrets_manager_secret_policy_get_command(client: AWSClient, args: Dict[str, Any]) -> CommandResults:
+    client = client.aws_session(
+        service=SERVICE,
+        role_arn=args.get('roleArn'),
+        role_session_name=args.get('roleSessionName'),
+        role_session_duration=args.get('roleSessionDuration'),
+    )
+
+    if not args.get('secret_id'):
+        return_error('secret_id is mandatory inorder to run this command!')
```
Comment (DeanArbel) [inline Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py:197]: Please replace this with a raise.

File: Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+        role_session_name=args.get('roleSessionName'),
+        role_session_duration=args.get('roleSessionDuration'),
+    )
+
+    creds_dict = dict()
+    if args.get('secret_id'):
+        try:
+            creds_dict[args.get('secret_id')] = client.get_secret_value(SecretId=args.get('secret_id'))
+        except Exception as e:
+            demisto.debug(f"Could not fetch credentials: {args.get('secret_id')}. Error: {e}")
```
Comment (DeanArbel) [inline Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py:228]: I believe in this case it's better to raise this as an error, as the user won't see this error.
@yuvalbenshalom wdyt?

File: Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+    readable_output = {'Name': response.get('Name', ''),
+                       'ARN': response.get('ARN', ''),
+                       'Policy': response.get('ResourcePolicy', '')}
+
+    human_readable = tableToMarkdown('AWS Secret Policy', readable_output)
+
+    return_results(CommandResults(
+        outputs_prefix='AWS.SecretsManager.Policy',
+        outputs=response,
+        readable_output=human_readable
```
Comment (DeanArbel) [inline Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py:212]: Please add `outputs_key_field`.

**Review by DeanArbel — CHANGES_REQUESTED** [review]

File: Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+
+    kwargs = dict()
+
+    if args.get('secret_id'):
+        kwargs['SecretId'] = args.get('secret_id')
+    if args.get('days_of_recovery'):
+        kwargs['RecoveryWindowInDays'] = int(args.get('days_of_recovery'))
+    if args.get('delete_immediately') is not None:
+        if args.get('days_of_recovery'):
+            return_error('Delete command cannot be executed with both args: delete_immediately and days_of_recovery')
```
[the remark this answers is not in the document]
Comment (ishaish103) [reply Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py:165]: done

File: Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+        kwargs['ForceDeleteWithoutRecovery'] = args.get('delete_immediately', False) == True
```
[the remark this answers is not in the document]
Comment (ishaish103) [reply Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py]: done

File: Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+def aws_secrets_manager_secret_restore_command(client: AWSClient, args: Dict[str, Any]) -> CommandResults:
+    client = client.aws_session(
+        service=SERVICE,
+        role_arn=args.get('roleArn'),
+        role_session_name=args.get('roleSessionName'),
+        role_session_duration=args.get('roleSessionDuration'),
+    )
+
+    if not args.get('secret_id'):
+        return_error('secret_id is mandatory inorder to run this command!')
```
[the remark this answers is not in the document]
Comment (ishaish103) [reply Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py]: done

File: Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py (hunk 1 above)
Comment (ishaish103) [reply Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py:97]: done

File: Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py (hunk 2 above)
Comment (ishaish103) [reply Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py:142]: done

File: Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py (hunk 3 above)
Comment (ishaish103) [reply Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py]: done

File: Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py (hunk 4 above)
Comment (ishaish103) [reply Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py]: done

File: Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py (hunk 5 above)
Comment (ishaish103) [reply Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py:198]: done

File: Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py (hunk 7 above)
Comment (ishaish103) [reply Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py:214]: done

**Comment by ShirleyDenkberg** [comment]
@ishaish103 
- Need descriptions for all outputs in the yml file.
- Need to see the generated README file.
- I created content for the pack README.
Please ping me when the descriptions and README are added so that I can review them.

**Comment by ShirleyDenkberg** [comment]
@ishaish103 Please generate an integration README file.
@DeanArbel @yuvalbenshalom

File: Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+            if not any(key in ["user", "password", "workgroup", "certificate"] for key in secret_as_dict.keys()):
+                credentials.append({
+                    "user": secret_as_dict.get("username", ""),
+                    "password": secret_as_dict.get("password", ""),
+                    "workgroup": secret_as_dict.get("workgroup", ""),
+                    "certificate": secret_as_dict.get("certificate", ""),
+                    "name": f'{creds_dict[cred_key].get("Name")}',
+                })
+            else:
+                LOG(f'({creds_dict[cred_key]}) has no keys supporting the format')
```
Comment (DeanArbel) [inline Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py:248]: Please replace this with `demisto.info` - `LOG` has been deprecated.

File: Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+        except Exception as e:
+            return_error(f'theres is a problem parsing ({creds_dict[cred_key]}) secret value, {e}')
```
Comment (DeanArbel) [inline Packs/Aws-SecretsManager/Integrations/AwsSecretsManager/AwsSecretsManager.py:250]: Please replace this with a raise - `return_error` is expected to run in `main` only.

**Review by DeanArbel — APPROVED** [review]
Great job

**Pull request merged**